include/aws/s3/private/s3_parallel_read_stream.h

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_S3_PARALLEL_READ_STREAM_H
+#define AWS_S3_PARALLEL_READ_STREAM_H
+
+#include <aws/s3/s3.h>
+
+#include <aws/common/ref_count.h>
+
+AWS_PUSH_SANE_WARNING_LEVEL
+
+struct aws_byte_buf;
+struct aws_future_bool;
+struct aws_input_stream;
+
+struct aws_event_loop_group;
+
+struct aws_parallel_input_stream {
+    const struct aws_parallel_input_stream_vtable *vtable;
+    struct aws_allocator *alloc;
+    struct aws_ref_count ref_count;
+
+    void *impl;
+};
+
+struct aws_parallel_input_stream_vtable {
+    /**
+     * Destroy the stream, its refcount has reached 0.
+     */
+    void (*destroy)(struct aws_parallel_input_stream *stream);
+
+    /**
+     * Read into the buffer in parallel.
+     * The implementation needs to support this to be invoked concurrently from multiple threads
+     *
+     * TODO: error handling:
+     * - The stream reach end before all parts
+     * - Failed to create a elg task to read
+     * - Asyc related stuff as the original async stream (If we cannot fill the dest, we wait?)
+     * - etc.
+     */
+    struct aws_future_bool *(*read)(
+        struct aws_parallel_input_stream *stream,
+        size_t start_position,
+        size_t end_position,
```

Comment:
uint64_t

Reply:
or offset/length is common in C apis

we don't need to worry about confusion with `object_range_end` which is confusingly an inclusive number

and we don't need to worry about start <= end

`uint64_t offset; size_t length`?